### ISSUE TITLE
Inconsistent inference from a rest parameter

### ISSUE THREAD

**Issue by cartant** [Microsoft/TypeScript#30824]
**TypeScript Version:**  3.5.0-dev.20190407 and 3.4 in TS playground

**Search Terms:** infer rest tuple

**Code**

```ts
type Action<T extends string = string> = { type: T };

declare function f<A extends Action>(...args: A[]): A[];
declare function g<A extends Action>(args: A[]): A[];

const A = "A";
const B = "B";
const C = "C";
declare const a: Action<typeof A>;
declare const b: Action<typeof B>;
declare const c: Action<typeof C>;

// Here b effects an error:
const r1 = f(a, b, c);

// Here r2 is inferred correctly:
const r2 = f(
    { type: A },
    { type: B },
    { type: C }
);

// Here r3 is inferred correctly:
const r3 = f(...[a, b, c]);

// Here r4 is inferred correctly:
const r4 = g([a, b, c]);

// Here r5 is inferred correctly:
const r5 = g([
    { type: A },
    { type: B },
    { type: C }
]);
```

**Expected behavior:**

In all of the calls to `f` and `g` in the example above, I would have expected no errors and the result to be inferred correctly.

**Actual behavior:**

Arguments passed to the function that accepts a rest parameter - `f` - work only if said arguments are literals or are spread from an array/tuple. If separate, declared variables are passed, an error is effected, stating that the arguments are expected to be of the same type.

Both variables and literals behave as expected with `g` - a function that accepts an array parameter, rather than a rest parameter.

**Playground Link:** 

[Here is the playground link](https://www.typescriptlang.org/play/#src=type%20Action%3CT%20extends%20string%20%3D%20string%3E%20%3D%20%7B%20type%3A%20T%20%7D%3B%0D%0A%0D%0Adeclare%20function%20f%3CA%20extends%20Action%3E(...args%3A%20A%5B%5D)%3A%20A%5B%5D%3B%0D%0Adeclare%20function%20g%3CA%20extends%20Action%3E(args%3A%20A%5B%5D)%3A%20A%5B%5D%3B%0D%0A%0D%0Aconst%20A%20%3D%20%22A%22%3B%0D%0Aconst%20B%20%3D%20%22B%22%3B%0D%0Aconst%20C%20%3D%20%22C%22%3B%0D%0Adeclare%20const%20a%3A%20Action%3Ctypeof%20A%3E%3B%0D%0Adeclare%20const%20b%3A%20Action%3Ctypeof%20B%3E%3B%0D%0Adeclare%20const%20c%3A%20Action%3Ctypeof%20C%3E%3B%0D%0A%0D%0A%2F%2F%20Here%20b%20effects%20an%20error%3A%0D%0Aconst%20r1%20%3D%20f(a%2C%20b%2C%20c)%3B%0D%0A%0D%0A%2F%2F%20Here%20r2%20is%20inferred%20correctly%3A%0D%0Aconst%20r2%20%3D%20f(%0D%0A%20%20%20%20%7B%20type%3A%20A%20%7D%2C%0D%0A%20%20%20%20%7B%20type%3A%20B%20%7D%2C%0D%0A%20%20%20%20%7B%20type%3A%20C%20%7D%0D%0A)%3B%0D%0A%0D%0A%2F%2F%20Here%20r3%20is%20inferred%20correctly%3A%0D%0Aconst%20r3%20%3D%20f(...%5Ba%2C%20b%2C%20c%5D)%3B%0D%0A%0D%0A%2F%2F%20Here%20r4%20is%20inferred%20correctly%3A%0D%0Aconst%20r4%20%3D%20g(%5Ba%2C%20b%2C%20c%5D)%3B%0D%0A%0D%0A%2F%2F%20Here%20r5%20is%20inferred%20correctly%3A%0D%0Aconst%20r5%20%3D%20g(%5B%0D%0A%20%20%20%20%7B%20type%3A%20A%20%7D%2C%0D%0A%20%20%20%20%7B%20type%3A%20B%20%7D%2C%0D%0A%20%20%20%20%7B%20type%3A%20C%20%7D%0D%0A%5D)%3B)

**Related Issues:** None found

**Comment by RyanCavanaugh** [comment]
This is an intentional difference.

Consider this code:
```ts
type Action<T extends string = string> = { type: T };
declare const a: Action<"A">, b: Action<"B">, c: Action<"C">;

declare function f<T extends Action>(...args: T[]): T[];
declare function g<T extends Action>(args: T[]): T[];
declare function h<T>(a: T, b: T, c: T): T[];

g([a, b, c]);
f(a, b, c); // ?
h(a, b, c);
```
`g` has to be OK because heterogeneous array literals are fine.

`h` should *clearly* be an error - otherwise generics don't mean anything (consider a function like `insert<T>(arr: T[], el: T): void` and calling `insert([1, 2, 3], "4")` - it'd be a disaster to infer `T = string|number`).

The question is whether it's more correct/consistent for `f` to behave like `g` or like `h` -- certainly you can have your own interpretation, but we chose that a rest arg array should behave more like an infinite set of formally-declared parameters than an array literal passed inline.

**Comment by cartant** [comment]
Thanks. BTW, the workaround appears to be [this](https://www.typescriptlang.org/play/#src=type%20Action%3CT%20extends%20string%20%3D%20string%3E%20%3D%20%7B%20type%3A%20T%20%7D%3B%0D%0A%0D%0Adeclare%20function%20f%3CA%20extends%20Action%2C%20R%20extends%20A%5B%5D%3E(...args%3A%20R)%3A%20R%3B%0D%0A%0D%0Aconst%20A%20%3D%20%22A%22%3B%0D%0Aconst%20B%20%3D%20%22B%22%3B%0D%0Aconst%20C%20%3D%20%22C%22%3B%0D%0Adeclare%20const%20a%3A%20Action%3Ctypeof%20A%3E%3B%0D%0Adeclare%20const%20b%3A%20Action%3Ctypeof%20B%3E%3B%0D%0Adeclare%20const%20c%3A%20Action%3Ctypeof%20C%3E%3B%0D%0A%0D%0Aconst%20r1%20%3D%20f(a%2C%20b%2C%20c)%3B).

**Comment by tjjfvi** [comment]
The above workaround can be simplified; the `A extends Action` generic in the other workaround is unused, as `A` is always inferred to be of type `Action<string>`. Only the `R` generic is usefully inferred, so the function signature can be reduced to:
```ts
// ...
declare function f<R extends Action[]>(...args: R): R;
// ...
```
[Playground Link](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=14&pc=1#code/C4TwDgpgBAggxsAlgewHYB4AqUIA9gSoAmAzlCcAE6KoDmUAvOVTbQHyNQDeUokAXFGwBfANwBYAFBSA9DKhEIcADYBDStABmAV1QIUqKJvQwc+QqVj60AGigAlMwWJkYAbQC6bABQA6f+q0JIL2AJQhEpKKKupautaGxo54zpbwSGiePv6+gcEO4Q6RUnBoFLCcAEQwlZGlqOUAQlWNtSVlwFAAwlVdbVFKahpQ9eWqgukG6HwQyJqwbJHRQ9CjnQBGEwnT4LPzjYtSy7EjHSNbGRgzc92H0pJrUJQAjJya3qp263ZwoaJAA)
